### PR TITLE
Bincompat hooks for revised hash/isEqual interop

### PR DESCRIPTION
**Pull request by tbkka**
This adds in hooks so that the new hash/isEqual interop (which bridges Obj-C hash/isEqual: calls to the corresponding Swift Hashable/Equatable conformances) can be selectively disabled based on the OS and/or client.

For now, enable the new semantics everywhere except Apple platforms (which have legacy apps that may be relying on the old semantics).

Resolves: rdar://121399248

### CODE REVIEW

**Comment by tbkka** [comment]
@swift-ci Please test

**Review by mikeash — APPROVED** [review]

**Comment by tbkka** [comment]
@swift-ci Please test

**Review by al45tair — APPROVED** [review]

File: stdlib/public/runtime/SwiftValue.mm [hunk 1]
```diff
@@ -458,6 +464,11 @@ - (NSUInteger)hash {
 	    selfHeader->type, hashableConformance);
   }
 
+  if (!runtime::bincompat::useLegacySwiftObjCHashing()) {
```
Comment (tbkka) [inline stdlib/public/runtime/SwiftValue.mm:467]: Darn.  That `!` should not have been there...   Follow-up PR will fix this.

Comment (mikeash) [reply stdlib/public/runtime/SwiftValue.mm:467]: Ouch. Lots of trouble from a line and a dot.

**Pull request merged**